azure/services/natgateways/client.go

```diff
@@ -90,15 +81,15 @@ func (ac *azureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.Resou
 	defer cancel()
 
 	pollOpts := &runtime.PollUntilDoneOptions{Frequency: asyncpoller.DefaultPollerFrequency}
-	result, err = poller.PollUntilDone(ctx, pollOpts)
+	resp, err := poller.PollUntilDone(ctx, pollOpts)
 	if err != nil {
 		// if an error occurs, return the poller.
 		// this means the long-running operation didn't finish in the specified timeout.
 		return nil, poller, err
 	}
 
 	// if the operation completed, return a nil poller
-	return result, nil, err
+	return resp.NatGateway, nil, err
```

Comment:
is there anything we can do to change the return type so it's not an `interface{}` to protect us from similar errors in the future?

Reply:
Each return value (both the `resp` and its `NatGateway` field) are individual structs without a common interface. https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork/v4@v4.0.0#NatGatewaysClientCreateOrUpdateResponse

Since this func (`CreateOrUpdateAsync`) is part of the `asyncpoller` interfaces, I'm not sure how we can avoid `interface{}` but I'm open to ideas.

Reply:
If all the returned resources (NatGateway, AgentPool, InboundNatRule, etc.) have some common field or fields, we can define an interface and return that. `ID`, `Name`, and `Type` are on the resources I've looked at.

Reply:
We could make an interface with the only two funcs that seem to be shared:

```go
// AzureResource is a basic interface for Azure resources.
type AzureResource interface {
	MarshalJSON() ([]byte, error)
	UnmarshalJSON(data []byte) error
}
```
Then we could return `AzureResource` instead of `interface{}`. It's more descriptive, and it would have prevented this particular bug.

But the `resp` part of `resp.NatGateway` also implements these funcs, so it wouldn't prevent accidentally returning `resp`. And we don't use the JSON funcs, so it feels a bit odd to use it just as a "marker" interface. WDYT?

Reply:
I don't have any better suggestions for now so let's keep it as-is but I'll try to keep thinking about this to see if we can improve it